nixos/modules/services/misc/omnom.nix

```
@@ -0,0 +1,281 @@
+              host = lib.mkOption {
+                type = lib.types.str;
+                default = "";
+                description = "SMTP server hostname.";
+              };
+              port = lib.mkOption {
+                type = lib.types.port;
+                default = 25;
+                description = "SMTP server port address.";
+              };
+              sender = lib.mkOption {
+                type = lib.types.str;
+                default = "Omnom <omnom@127.0.0.1>";
+                description = "Omnom sender e-mail.";
+              };
+              send_timeout = lib.mkOption {
+                type = lib.types.int;
+                default = 10;
+                description = "Send timeout duration in seconds.";
+              };
+              connection_timeout = lib.mkOption {
+                type = lib.types.int;
+                default = 5;
+                description = "Connection timeout duration in seconds.";
+              };
+            };
+          };
+        };
+        default = { };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !lib.hasAttr "username" cfg.settings.smtp;
+        message = ''
+          `services.omnom.settings.smtp.username` must be defined in `services.omnom.secretsFile`.
+        '';
+      }
+      {
+        assertion = !lib.hasAttr "password" cfg.settings.smtp;
+        message = ''
+          `services.omnom.settings.smtp.password` must be defined in `services.omnom.secretsFile`.
+        '';
+      }
+      {
+        assertion = !(cfg.settings.storage.root != "${cfg.dataDir}/static/data");
+        message = ''
+          For Omnom to access the snapshots, it needs the storage root
+          directory to be inside the service's working directory.
+
+          As such, `services.omnom.settings.storage.root` must be the same as
+          `''${services.omnom.dataDir}/static/data`.
+        '';
+      }
+    ];
+
+    environment.etc."omnom/config.yml".source = configFile;
+
+    systemd.services.omnom = {
+      path = with pkgs; [
+        yq-go # needed by startup script
+      ];
+
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = "omnom";
+        WorkingDirectory = cfg.dataDir;
+        Restart = "on-failure";
+        RestartSec = "10s";
+        LoadCredential = lib.mkIf (cfg.secretsFile != null) [
+          "SECRETS_FILE:${cfg.secretsFile}"
+        ];
+      };
+      script = ''
+        export CONFIG_FILE=${configFile}
+
+        ${lib.optionalString (cfg.secretsFile != null) ''
+          # merge secrets into main config
+          yq eval-all "select(fileIndex == 0) * select(fileIndex == 1)" \
+            "${configFile}" "$CREDENTIALS_DIRECTORY/SECRETS_FILE" > "${cfg.dataDir}/config.yml"
+
+          CONFIG_FILE="${cfg.dataDir}/config.yml"
+        ''}
+
+        ${lib.getExe cfg.package} listen --config "$CONFIG_FILE"
```

```suggestion
        install -m 600 ${configFile} $STATE_DIRECTORY/config.yml
        
        ${lib.optionalString (cfg.smtpPasswordFile != null) ''
          # merge password into main config 
          yq -i "smtp.password = load(env(CREDENTIALS_DIRECTORY) + "SECRETS_FILE") \
            "$STATE_DIRECTORY/config.yml"
        ''}

        ${lib.getExe cfg.package} listen --config "$STATE_DIRECTORY/config.yml"
```

Let's always keep the config into `/var/lib/omnom` otherwise things are going to get confusing :)

I've also modified the `yq` expr to only insert the SMTP password, this also make the secret management a bit simpler as folks only have to have the password in the file not a whole yaml structure.

---

I haven't tested that the `yq` expr does what I expect it to do. If you could give it a check that would be great :)

---

> I haven't tested that the yq expr does what I expect it to do. If you could give it a check that would be great :)

I had to modify it a bit, but I tested it and it works. I've also changed the option's name from `secretsFile` to `passwordFile` since that's more descriptive of what we're storing, now.